app/services/ping.rb

```diff
@@ -0,0 +1,65 @@
+class Ping
+  class << self
+    def ping
+      {
+        'foreman': {
+          database: ping_database
+        }
+      }.merge(plugins(:ping))
+    end
+
+    def statuses
+      {
+        'foreman': {
+          version: SETTINGS[:version].full,
+          api: {
+            version: 2
+          },
+          plugins: Foreman::Plugin.all,
+          smart_proxies: statuses_smart_proxies,
+          compute_resources: statuses_compute_resources
+        }
+      }.merge(plugins(:status))
+    end
+
+    private
+
+    def ping_database
+      ActiveRecord::Base.connection.active?
+    end
+
+    def statuses_smart_proxies
+      response = []
+      SmartProxy.all.includes(:features).each do |proxy|
+        response << {
+          name: proxy.name,
+          version: proxy.statuses[:version].version['version'],
+          features: proxy.statuses[:version].version['modules'],
+          failed_features: proxy.statuses[:logs].logs.failed_modules
+        }
+      end
+      response
+    end
+
+    def statuses_compute_resources
+      response = []
+      ComputeResource.all.index.each do |resource|
+        errors = resource.ping
+        response << {
+          name: resource.name,
+          status: errors.empty? ? 'ok' : 'FAIL',
+          errors: errors.full_messages
+        }
+      end
+      response
+    end
+
+    def plugins(action)
+      Foreman::Plugin.all.inject({}) do |all, plugin|
+        next all if plugin.ping_extension.nil?
+
+        all.update({ "#{plugin.name}": plugin.ping_extension.send(action) })
```

Comment:
This is a very fuzzy plugin API, I'd vote for something more formal like ping_callback and status_callback blocks/procs.